src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,171 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
             return merged;
         }
 
+        /// <summary>
+        /// Merges the new change ranges into the old change ranges, adjusting the new ranges to be with respect to the original text
+        /// (with neither old or new changes applied) instead of with respect to the original text after "old changes" are applied.
+        ///
+        /// This may require splitting, concatenation, etc. of individual change ranges.
+        /// </summary>
         private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRange> oldChanges, ImmutableArray<TextChangeRange> newChanges)
```

Comment:
Is there an implicit contract that both arrays here are ordered in some way? If so can we add that to the remarks section of the doc comment?

Reply:
Yes, the logic below relies on this ordering assumption. Maybe we could add assertions too.

---
In reply to: [501910016](https://github.com/dotnet/roslyn/pull/48420#discussion_r501910016) [](ancestors = 501910016)

Reply:
The `add` local function will assert when spans are overlapping or out of order.

Reply:
can we contract-throw instead?  asserts don't help us in retail builds.

Note: i have a PR that adds some asserts at an earlier point in time to also catch bad info getting stored into the heap: https://github.com/dotnet/roslyn/pull/48474/files

Reply:
Done